arrabiata/src/witness.rs

```diff
@@ -0,0 +1,1098 @@
+    // at the end of the whole execution
+    // FIXME: use a blinded comm and also fold the blinder
+    pub ivc_accumulator_e1: Vec<PolyComm<E1>>,
+
+    // FIXME: use a blinded comm and also fold the blinder
+    pub ivc_accumulator_e2: Vec<PolyComm<E2>>,
+
+    /// Commitments to the previous instances
+    pub previous_commitments_e1: Vec<PolyComm<E1>>,
+    pub previous_commitments_e2: Vec<PolyComm<E2>>,
+    // ----------------
+
+    // ----------------
+    // Data only used by the interpreter while building the witness over time
+    /// The index of the latest allocated variable in the circuit.
+    /// It is used to allocate new variables without having to keep track of the
+    /// position.
+    pub idx_var: usize,
+
+    /// The index of the latest allocated public inputs in the circuit.
+    /// It is used to allocate new public inputs without having to keep track of
+    /// the position.
+    pub idx_var_pi: usize,
+
+    /// Current processing row. Used to build the witness.
+    pub current_row: usize,
+
+    /// State of the current row in the execution trace
+    pub state: [BigInt; NUMBER_OF_COLUMNS],
+
+    /// Contain the public state
+    // FIXME: I don't like this design. Feel free to suggest a better solution
+    pub public_state: [BigInt; NUMBER_OF_PUBLIC_INPUTS],
+
+    /// Selectors to activate the gadgets.
+    /// The size of the outer vector must be equal to the number of gadgets in
+    /// the circuit.
+    /// The size of the inner vector must be equal to the number of rows in
+    /// the circuit.
+    ///
+    /// The layout columns/rows is used to avoid rebuilding the arrays per
+    /// column when committing to the witness.
+    pub selectors: Vec<Vec<bool>>,
+
+    /// While folding, we must keep track of the challenges the verifier would
+    /// have sent in the SNARK, and we must aggregate them.
+    // FIXME: nothing is done yet, and the challenges haven't been decided yet.
+    // See top-level documentation of the interpreter for more information.
+    pub challenges: Vec<BigInt>,
+
+    /// Keep the current executed instruction
+    /// This can be used to identify which gadget the interpreter is currently
+    /// building.
+    pub current_instruction: Instruction,
+
+    /// The sponges will be used to simulate the verifier messages, and will
+    /// also be used to verify the consistency of the computation by hashing the
+    /// public IO.
+    // IMPROVEME: use a list of BigInt? It might be faster as the CPU will
+    // already have in its cache the values, and we can use a flat array
+    pub sponge_e1: [BigInt; POSEIDON_STATE_SIZE],
+    pub sponge_e2: [BigInt; POSEIDON_STATE_SIZE],
+
+    /// List of public inputs, used first to verify the consistency of the
+    /// previous iteration.
+    pub current_iteration: u64,
+
+    /// A previous hash, encoded in 2 chunks of 128 bits.
+    pub previous_hash: [u128; 2],
```

Comment:
why not a field element

Reply:
Because it has to be instantiated for both fields. We might want to parametrize later. We should bench if it is faster or not than bigint.